rust/connlib/tunnel/src/tests.rs

```diff
@@ -590,27 +599,68 @@ impl ReferenceStateMachine for ReferenceState {
                     }
                     Some(_) | None => {
-                        // Depending on the DNS query type, we filter the resolved addresses.
-                        let ips_resolved_by_query = all_ips.iter().copied().filter({
-                            #[allow(clippy::wildcard_enum_match_arm)]
-                            match r_type {
-                                RecordType::A => {
-                                    &(|ip: &IpAddr| ip.is_ipv4()) as &dyn Fn(&IpAddr) -> bool
-                                }
-                                RecordType::AAAA => {
-                                    &(|ip: &IpAddr| ip.is_ipv6()) as &dyn Fn(&IpAddr) -> bool
+                        // If we are NOT connected to this resource yet, we will connect and expire existing DNS records in the process.
+                        if !state.client_dns_records.contains_key(&domain) {
+                            for ip in &all_ips {
+                                for (prev_domain, prev_ips) in state.client_dns_records.iter_mut() {
+                                    prev_ips.retain(|(prev_ip, _)| {
+                                        if prev_ip == ip {
+                                            tracing::debug!(
+                                                "Expiring DNS record: '{prev_domain} -> {prev_ip}'"
+                                            );
+                                            return false;
+                                        }
+
+                                        true
+                                    });
                                 }
-                                _ => unimplemented!(),
                             }
-                        });
 
-                        state
-                            .client_dns_records
-                            .entry(domain.clone())
-                            .or_default()
-                            .extend(ips_resolved_by_query);
+                            // We only add new records when we first connect to a resource.
+                            for ip in all_ips {
+                                state
+                                .client_dns_records
+                                .entry(domain.clone())
+                                .or_default()
+                                .push({
+                                    let ip_status = {
+                                        #[allow(clippy::wildcard_enum_match_arm)]
+                                        match r_type {
+                                            RecordType::A if ip.is_ipv4() => IpStatus::Queried,
+                                            RecordType::AAAA if ip.is_ipv6() => IpStatus::Queried,
+                                            RecordType::A | RecordType::AAAA => IpStatus::Internal,
+                                            _ => unimplemented!(),
+                                        }
```

Comment:
I'm finding these match blocks hard to read.
I think `RecordType::A | RecordType::AAAA` threw me off. At first I thought it meant, "If the DNS query is for both A and AAAA" but it's actually part of the match syntax, it means, "If the DNS query is for A or for AAAA", which catches IPv6 addresses for A queries and IPv4 addresses for AAAA queries.

I guess a DNS query cannot be for both A and AAAA records, but the server is allowed to respond with both records to save a round-trip?